### ISSUE TITLE
PSYKE: Generate statements for writing out the LFRic data

### ISSUE THREAD

**Issue by TeranIvy** [stfc/PSyclone#234]
There are several aspects of kernel extraction: 

1) Marking nodes for extraction, 
2) Generating the driver layer to run extracted loops with kernel calls and possible optimisations, 
3) Generating the statements in the LFRic code for writing out LFRic data required to run the kernel driver,
4) Generating the read statements and argument declarations in the kernel driver.

The first and the basic elements of the second point are being addressed in #139 and associated PR #218. This issue will deal with the third point by using the following resources:

- Extending the functionality initially developed in #7,
- Examples in the [LFRic-microbenchmarks](https://github.com/christophermaynard/LFRic-microbenchmarks) repository,
- [dino_mod library in LFRic](https://code.metoffice.gov.uk/trac/lfric/browser/LFRic/trunk/infrastructure/source/PSYKE/dino_mod.F90) which the generated statements will be using.

The functionality developed in this issue will also contribute to points 2 and 4 (populating the driver layer). A separate issue will be required to round off the kernel driver (horizontal loop bounds, declaration of variables etc.).

**Comment by TeranIvy** [comment]
Added @christophermaynard, @rupertford, @arporter  and @samcusworth to the assignees.

**Comment by TeranIvy** [comment]
- Example of generated LFRic PSy layer with hand-written dump statements for the [CMA_appinv microbenchmark](https://github.com/christophermaynard/LFRic-microbenchmarks/tree/master/CMA_appinv) is [gw_pressure_precon_alg_mod_psy.f90.txt](https://github.com/stfc/PSyclone/files/2558250/gw_pressure_precon_alg_mod_psy.f90.txt)
- Example of generated LFRic PSy layer with hand-written dump statements for the  [LMA microbenchmark](https://github.com/christophermaynard/LFRic-microbenchmarks/tree/master/LMA) is [helmholtz_solver_alg_mod_psy.f90.txt](https://github.com/stfc/PSyclone/files/2558251/helmholtz_solver_alg_mod_psy.f90.txt)

The LFRic PSy layer examples and the kernel drivers are generated (or written) using the specific PSyclone settings for the LFRic code from `global.py` script. These  are `Dynamo0p3ColourTrans`, `Dynamo0p3OMPLoopTrans` and `OMPParallelTrans`. The upper loop bounds in LFRic PSy examples indicate that distributed memory is used as well:
```
      DO colour=1,mesh%get_ncolours()
        !$omp parallel default(shared), private(cell)
        !$omp do schedule(static)
        DO cell=1,mesh%get_last_halo_cell_per_colour(colour,1)
```
whereas the upper loop bounds in the kernel driver examples are almost the same as in the code generated without the distributed memory:
```
      DO colour=1,ncolour
        !$omp parallel default(shared), private(cell)
        !$omp do schedule(static)
        DO cell=1,ncp_colour(colour)
```

**Comment by hiker** [comment]
Quick question: why is distributed-memory disabled? Shouldn't it be possible to:
- give each process a unique filename (by adding the rank)
- write the data independently for each process
- then we run the driver in parallel, but independently (i.e. no communication in the driver, but the data for each process would be executed at the same time)?

**Comment by arporter** [comment]
I think the problem is mainly due to the way that the set_dirty/set_clean calls are created - because they are generated by the loop object there's no easy way of excluding them from an extraction region. @teranivy or @rupertford might know more...

**Comment by TeranIvy** [comment]
@arporter is right, `set_dirty` calls are part of the `gen_code` method in `DynLoop` and would always end up in extract region. @hiker, if you have a way how to avoid the issue I'd be very happy to hear it :smiley:

**Comment by rupertford** [comment]
I agree, it would be good to support parallel.

In terms of `set_dirty()` and `set_clean()` there is an issue related to this ... [search, grep, find, recurse, ahah] ... #450, not that this helps :-). I guess a hacky solution would be to switch off the generation of these calls when extracting?

The other issue is that we wanted to avoid having infrastructure calls within an extract regions and we could end up with halo exchange calls and global sums in theory (but we could (and may already) just disallow these in regions so it would just be a restriction).

**Comment by arporter** [comment]
In fact, I suspect this can be closed if @hiker and @TeranIvy agree?

**Comment by hiker** [comment]
@arporter , I am ok to close this. The halo exchange and setdirty/clean calls are a bit of an issue for the driver creation (I manually removed these), but that's indeed a different issue, for writing data it doesn't matter much. Limited support for MPI should be possible by adding the rank to the created file names - after which each rank could be run independently (or with a trivial mpi wrapper). Including MPI communication would be very difficult :(

**Comment by hiker** [comment]
At this stage it appears that `set_dirty()` calls are included in the region, but `halo_exchange` are not. Without digging into details that might just depend on which of the corresponding nodes exist at the time the extract transformation is done.
But I actually don't worry about this at all (only exception would be for profiling, to avoid measuring the halo exchange cost), nothing in the extract transformation is affected by that (all variables that have `set_dirty()` etc called, will already be extracted because of the kernel meta info.

The only issue will be the driver creation, but that's a different issue (and trivial solution is actually to temporarily switch off distributed memory while writing out the driver). So I am still in favour of closing this ticket
